rollout/trafficrouting/openshift/openshift.go

```diff
@@ -0,0 +1,177 @@
+type Reconciler struct {
+	Cfg ReconcilerConfig
+	Log *logrus.Entry
+}
+
+// NewReconciler will build and return an OpenShift Reconciler
+func NewReconciler(cfg ReconcilerConfig) *Reconciler {
+	return &Reconciler{
+		Cfg: cfg,
+		Log: logutil.WithRollout(cfg.Rollout),
+	}
+}
+
+// SetWeight changes the route configuration according to desiredWeight
+func (r *Reconciler) SetWeight(desiredWeight int32) error {
+	r.sendNormalEvent("WeightUpdate", fmt.Sprintf("Set weight to %d", desiredWeight))
+	ctx := context.TODO()
+	routeNameList := r.Cfg.Rollout.Spec.Strategy.Canary.TrafficRouting.Openshift.Routes
+	doneChan := make(chan struct{})
+	errChan := make(chan error)
+	defer close(errChan)
+	var wg sync.WaitGroup
+	wg.Add(len(routeNameList))
+	for _, routeName := range routeNameList {
+		go func(routeName string) {
+			defer wg.Done()
+			err := r.updateRoute(ctx, routeName, desiredWeight)
+			if err != nil {
+				errChan <- err
+			}
+		}(routeName)
+	}
+	go func() {
+		wg.Wait()
+		close(doneChan)
+	}()
+
+	errs := []error{}
+	done := false
+	for !done {
+		select {
+		case <-doneChan:
+			done = true
+		case err := <-errChan:
+			errs = append(errs, err)
+		}
+	}
+	return formatErrors(errs)
+}
+
+func formatErrors(errs []error) error {
+	errorsCount := len(errs)
+	if errorsCount == 0 {
+		return nil
+	} else if errorsCount == 1 {
+		return errs[0]
+	}
+	var errMsg strings.Builder
+	errMsg.WriteString(fmt.Sprintf("%d errors found: ", errorsCount))
+	for _, err := range errs {
+		errMsg.WriteString(fmt.Sprintf("[%s]", err))
+	}
+	return errors.New(errMsg.String())
+}
+
+func (r *Reconciler) getRoute(ctx context.Context, routeName string) (*routev1.Route, error) {
+	return r.Cfg.Client.RouteV1().Routes(r.Cfg.Rollout.GetNamespace()).Get(ctx, routeName, metav1.GetOptions{})
+}
+
+//Update default backend weight,
+//remove alternateBackends if weight is 0,
+//otherwise update alternateBackends
+func (r *Reconciler) updateRoute(ctx context.Context, routeName string, desiredWeight int32) error {
+	// get the route
+	route, err := r.getRoute(ctx, routeName)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			msg := fmt.Sprintf("Route %q not found", routeName)
+			r.sendWarningEvent("OpenshiftRouteNotFound", msg)
+		}
+		return err
+	}
+
+	// update default backend weight
+	altWeight := 100 - desiredWeight
+	r.Log.Infof("updating default backend weight to %d", altWeight)
+	route.Spec.To.Weight = &altWeight
+	if desiredWeight == 0 {
+		r.Log.Infof("deleting alternateBackends")
+		route.Spec.AlternateBackends = nil
+	} else {
+		r.Log.Infof("updating alternate backend weight to %d", desiredWeight)
+		route.Spec.AlternateBackends = []routev1.RouteTargetReference{{
+			Kind:   "Service",
+			Name:   r.Cfg.Rollout.Spec.Strategy.Canary.CanaryService,
+			Weight: &desiredWeight,
+		}}
+	}
+
+	_, err = r.Cfg.Client.RouteV1().Routes(r.Cfg.Rollout.GetNamespace()).Update(ctx, route, metav1.UpdateOptions{})
+
+	return err
+}
```

Comment:
The way this code is currently written is that it will perform an update operation on the Route object even if there are no changes to be made, and the route backends/weights are already what we desire. Could we make this function to be a no-op and avoid the K8s API call altogether if the weights are already as expected? 

Also, we should emit a K8s event to indicate when a modification of the OpenShift route actually took place (so that when debugging, it is clear in the logs when route objects are modified).

You can see examples of this in the istio implementation:
```
		r.recorder.Eventf(r.rollout, record.EventOptions{EventReason: "UpdatedVirtualService"}, "VirtualService `%s` set to desiredWeight '%d'", vsvcName, desiredWeight)
```